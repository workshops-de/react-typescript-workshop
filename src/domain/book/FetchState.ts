export type FetchState = "initial" | "loading" | "success" | "error";
