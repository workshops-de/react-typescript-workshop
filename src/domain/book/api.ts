import { Book } from "./Book";

export const fetchBooks = async () => {
  const response = await fetch("http://localhost:4730/books");
  const data = await response.json();
  return data as Book[];
};

export const fetchBook = async (isbn: string) => {
  const response = await fetch("http://localhost:4730/books/" + isbn);
  const data = await response.json();
  return data as Book;
};
