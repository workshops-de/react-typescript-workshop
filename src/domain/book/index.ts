export * from "./api";
export * from "./Book";
export * from "./exampleBooks";
export * from "./hooks";
