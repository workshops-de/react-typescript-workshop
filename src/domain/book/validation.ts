import { z } from "zod";

export const validateBookForm = z
  .object({
    title: z.string().min(5),
    author: z.string().min(2),
    numPages: z.coerce.number().min(10),
  })
  .refine(
    ({ title, numPages }) => {
      if (title === "Bible") {
        return numPages > 1000;
      }

      return true;
    },
    { message: "The Bible must have more than 1000 pages.", path: ["numPages"] }
  );
