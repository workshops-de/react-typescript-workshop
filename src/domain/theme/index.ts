export * from "./Theme";
export * from "./ThemeContext";
