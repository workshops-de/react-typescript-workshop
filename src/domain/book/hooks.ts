import { useState, useEffect } from "react";
import { Book } from "./Book";
import { fetchBooks } from "./api";

export const useBooks = () => {
  const [books, setBooks] = useState<Book[]>([]);
  useEffect(() => {
    // Possibility 1: local async function that gets immediately called
    async function fetchData() {
      setBooks(await fetchBooks());
    }
    fetchData();

    // Possibility 2: respond to promise resolution with .then handler
    // fetchBooks().then(setBooks);
  }, []);

  return {
    books,
  };
};
