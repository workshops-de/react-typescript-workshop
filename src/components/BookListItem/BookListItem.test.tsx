import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import { BookListItem } from ".";
import { Book } from "../../domain/book";

describe("components/BookListItem", () => {
  const book = {
    id: "1",
    title: "My book",
    isbn: "my-isbn",
    price: "$11.11",
  } as Book;

  test("renders book with title", () => {
    render(<BookListItem book={book} />, {
      wrapper: MemoryRouter,
    });

    expect(screen.getByText(book.title)).toBeTruthy();
  });

  test(`renders a booklist item correctly`, () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <BookListItem book={book} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
