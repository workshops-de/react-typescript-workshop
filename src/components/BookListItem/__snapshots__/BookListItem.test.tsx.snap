// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BookListItem renders a booklist item correctly 1`] = `
<div>
  <a
    className=""
    href="/books/my-isbn"
    onClick={[Function]}
  >
    <h2
      style={
        Object {
          "textDecoration": "none",
        }
      }
    >
      My book
    </h2>
  </a>
  <h3 />
  <div>
    by 
    <strong />
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      👏 
      0
    </button>
  </div>
  <button
    onClick={[Function]}
  >
    +
  </button>
</div>
`;
